### ISSUE TITLE
Auto-populate node labels with node information from cloud provider

### ISSUE THREAD

**Issue by davidopp** [kubernetes/kubernetes#9044]
and add scheduling policies that allow things like spreading across failure domains (to the extent that such information is available from the cloud provider).

**Comment by roberthbailey** [comment]
This seems like a cloud provider specific ask. What label would you suggest for GCE (as an example)? Or AWS? I could see how bare metal deployment might be interested in encoding which rack each node is on, but there is also the option of applying labels to nodes after the fact. e.g. if you named your machines rack1-machine1..N you could run a script that iterated over each node, split the node name into two parts and applied a label for the rack to each node.

**Comment by davidopp** [comment]
For bare metal nodes could specify these labels when they register, as you did with hostname in #6949 (of course this begs the question how the nodes find out which labels they should register way). Or, as you say, they could come up with no labels and the user could add the labels after the fact through the API server. Labels could include things like rack, CPU type (in the case of heterogeneous cluster), and any special features or behaviors that aren't explicitly represented by a scheduler resource.

For cloud providers, I agree this feature is definitely provider-specific, and probably less useful, since the information may not be exposed to the user. But you could imagine propagating whatever information is available in the VM up to the k8s node object -- for example, in the GCE case things like VM attributes, tags, and zone. Zone would allow the scheduler to spread services across zones in the case of a cluster that spans a region.

**Comment by bgrant0607** [comment]
Internally, nodes automatically surface a variety of attributes that can be used in constraints. We should do so also, and could also use them for spreading. The scheduler can be configured with which labels to use for spreading.

**Comment by davidopp** [comment]
Related comment from @bgrant0607 here
https://github.com/kubernetes/kubernetes/issues/11470#issuecomment-124283056

> we'll need a way for Kubelets to export
>
> node attributes, represented as labels
> node resources, represented as non-zero capacity of generic counted resources

**Comment by davidopp** [comment]
Also, node hostname, so people can use label selectors instead of setting nodeName directly.

**Comment by bgrant0607** [comment]
Re. nodeName: see #12545.

**Comment by bgrant0607** [comment]
See also #12090

**Comment by davidopp** [comment]
Some example node properties that are automatically propagated into node attributes in Nomad:
https://www.nomadproject.io/docs/jobspec/index.html
(search for "Below is a table documenting common node attributes")

**Comment by davidopp** [comment]
And the ones that Swarm automatically populates are on this page
https://docs.docker.com/swarm/scheduler/filter/

**Comment by bgrant0607** [comment]
Zone/region labels were added by #16866

**Comment by bgrant0607** [comment]
See also: #13524, #13661, #16090

**Comment by fabioy** [comment]
This feels like we're going to start having a whole hodgepodge of these labels, with somewhat poorly defined schema. Perhaps we should consider augmenting labels with things like prefixes/selectors and defining a schema for them? (Or come up with other mechanisms, like extensible metadata?).

**Comment by mikedanese** [comment]
@fabioy we have api conventions for label/annotation key prefixes. Does that mitigate your concern (assuming we follow them)? See https://github.com/mikedanese/kubernetes/blob/master/docs/devel/api-conventions.md#label-selector-and-annotation-conventions (last paragraph).

**Comment by resouer** [comment]
OK, after reading thru the context. What I want is automatically adding label from configure flags of `kubelet` & `api-server`, like `privileged=true`, `hostnetwork="*"`, which does not need cloud provider's help.

**Comment by davidopp** [comment]
> OK, after reading thru the context. I can say what I want is automatically adding label from configure flags of kubelet & api-server, like privileged=true, hostnetwork="*", which does not need cloud provider's help.

Let's keep all of these use cases in this issue, rather than creating separate issues, for sanity.

**Comment by resouer** [comment]
@davidopp Yes, of course.

Do you think it make sense that I simply add labels [here]( https://github.com/kubernetes/kubernetes/pull/6949/files#diff-bf28da68f62a8df6e99e447c4351122dR692), if we can narrow down what we need to support?

Another question is if it is optional by default?

**Comment by mikedanese** [comment]
You'd want the labels to be updated if they changed/

**Comment by davidopp** [comment]
@resouer You should read through all of the issues that are mentioned in this issue, as they discuss possible approaches how to do it.

**Comment by davidopp** [comment]
Feature complete deadline for 1.2 is Friday, looks like we're not going to get to this, which is unfortunate, but I think we have no choice but to bounce it to the next release, unless someone else has a comment or suggestion...

**Comment by therc** [comment]
Another use case: until there's anything like actual GPU resource scheduling, getting the AWS instance type can help schedule pods only on machines with GPUs. It might also be useful for debugging purposes. I have cycles to help with the code, if it's likely to get merged. (We do need something like this and can live with 1.2+patches if it's too late for the release.)

**Comment by davidopp** [comment]
We have some standard labels emerging, and we've been putting them in
`pkg/api/unversioned/well_known_labels.go`

I think we will eventually have cloud-provider-specific node labels, e.g. for features that are available on one cloud provider but not others, so we should think about a label key namespacing scheme for that. It's a little unfortunate that our label keys are all top-level now, since it seems the ideal scheme would be something like

```
kubernetes.io/generic/hostname
kubernetes.io/generic/instance-type
kubernetes.io/gce/foo
kubernetes.io/aws/bar
```

I guess we can still do this, but the generic labels wouldn't have the `generic` part of the path (to be clear, those labels already exist). Or we could move to the scheme I'm proposing but still leave `kubernetes.io/hostname` and `kubernetes.io/instance-type` around for backward compatibility.

**Comment by pnovotnak** [comment]
It would be nice if nodes were labeled with their configured resource access (ie; compute-rw, storage-rw, etc.)

I had an issue yesterday where I created a Container Engine pool via the Cloud Console and it didn't have the correct scopes (which is obv a separate issue). My pods were being scheduled but the applications would crash when they tried to access a specific resource. It was not immediately obvious from the logs what was going wrong. Though those applications should have had more informative logging, it would be nice to be able to write a node selector to explicitly prohibit pods from being scheduled where they do not have requisite resource access.

**Comment by roberthbailey** [comment]
>  it would be nice to be able to write a node selector to explicitly prohibit pods from being scheduled where they do not have requisite resource access.

You can do this today if you mentally map the scopes necessary to the name of the node pool (which appears as a label on the pool). 

e.g. Create a pool named "nodes-with-compute-rw". Then pods that need the compute-rw scope can be restricted to nodes within that pool using a selector. 

Obviously this isn't automatic and the scopes aren't as discoverable as you'd like, but it may help a bit.

**Comment by justinsb** [comment]
@davidopp we do have failure-domain labels now; can you clarify what else needs to be done here before we can close this issue?

**Comment by davidopp** [comment]
Yeah I guess this issue is too vague to be useful at this point, we can close it.
